src/lib.rs

```diff
@@ -0,0 +1,155 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use core::cell::UnsafeCell;
+
+pub mod impls;
+
+pub trait ConstInit {
+    /// Create a new instance.
+    ///
+    /// This is a const instead of a method to allow creating instances in const context.
+    const INIT: Self;
+}
+
+/// Raw mutex trait.
+///
+/// This mutex is "raw", which means it does not actually contain the protected data, it
+/// just implements the mutex mechanism. For most uses you should use [`BlockingMutex`]
+/// instead, which is generic over a ScopedRawMutex and contains the protected data.
+///
+/// # Safety
+///
+/// ScopedRawMutex implementations must ensure that, while locked, no other thread can lock
+/// the RawMutex concurrently. This can usually be implemented using an [`AtomicBool`]
+/// to track the "taken" state. See [crate::impls] for examples of correct implementations.
+///
+/// Unsafe code is allowed to rely on this fact, so incorrect implementations will cause undefined behavior.
+///
+/// [`AtomicBool`]: core::sync::atomic::AtomicBool
+pub unsafe trait ScopedRawMutex {
+    /// Lock this `ScopedRawMutex`, calling `f()` after the lock has been acquired, and releasing
+    /// the lock after the completion of `f()`.
+    ///
+    /// If this was successful, `Some(R)` will be returned. If the mutex was already locked,
+    /// `None` will be returned
+    #[must_use]
+    fn try_lock<R>(&self, f: impl FnOnce() -> R) -> Option<R>;
+
+    /// Lock this `ScopedRawMutex`, calling `f()` after the lock has been acquired, and releasing
+    /// the lock after the completion of `f()`.
+    ///
+    /// Implementors may choose whether to block or panic if the lock is already locked.
+    /// It is recommended to panic if it is possible to know that deadlock has occurred.
+    ///
+    /// For implementations on a system with threads, blocking may be the correct choice.
+    ///
+    /// For implementations where a single thread is present, panicking immediately may be
+    /// the correct choice.
+    fn lock<R>(&self, f: impl FnOnce() -> R) -> R;
```

Comment:
I still have a preference for names that are different from the names that a RAII mutex would use, so that downstream synchronization primitive impls can use the same names as the raw mutex for their API methods while _also_ providing a RAII interface when the underlying mutex supports it (e.g. when it also implements `lock_api`'s traits). IMHO these should be `with_lock`/`try_with_lock`, or `locked`/`try_locked`.